trustfall_core/src/ir/ty.rs

```diff
@@ -0,0 +1,203 @@
+    ///
+    /// let nullable = false;
+    /// let ty = Type::new_named_type("String", nullable);
+    ///
+    /// assert_eq!(ty.to_string(), "String!");
+    /// assert_eq!(ty, Type::new("String!").unwrap());
+    /// ```
+    pub fn new_named_type(base_type_name: &str, nullable: bool) -> Type {
+        Type { ty: GQLType { base: BaseType::Named(Name::new(base_type_name)), nullable } }
+    }
+
+    /// Creates a new list [`Type`] from an individual [`Type`].
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let inner_nullable = false;
+    /// let inner_ty = Type::new_named_type("String", inner_nullable);
+    ///
+    /// let outer_nullable = true;
+    /// let ty = Type::new_list_type(inner_ty, outer_nullable);
+    ///
+    /// assert_eq!(ty.to_string(), "[String!]");
+    /// assert_eq!(ty, Type::new("[String!]").unwrap());
+    /// ```
+    pub fn new_list_type(inner_type: Type, nullable: bool) -> Type {
+        Type { ty: GQLType { base: BaseType::List(Box::new(inner_type.ty)), nullable } }
+    }
+
+    /// Returns a new type that is the same as this one, but with the passed nullability.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let nullable_ty = Type::new("Int").unwrap();
+    /// assert_eq!(nullable_ty.is_nullable(), true);
+    /// let non_nullable_ty = nullable_ty.with_nullability(false);
+    /// assert_eq!(non_nullable_ty.is_nullable(), false);
+    ///
+    /// // The original type is unchanged.
+    /// assert_eq!(nullable_ty.is_nullable(), true);
+    /// ```
+    pub fn with_nullability(&self, nullable: bool) -> Type {
+        Type { ty: GQLType { base: self.ty.base.clone(), nullable } }
+    }
+
+    /// Returns whether this type is nullable, at the top level, see example.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let nullable_ty = Type::new("[Int!]").unwrap();
+    /// assert_eq!(nullable_ty.is_nullable(), true); // the list is nullable
+    ///
+    /// let nullable_ty = Type::new("Int!").unwrap();
+    /// assert_eq!(nullable_ty.is_nullable(), false); // the `Int` is nonnullable
+    /// ```
+    pub fn is_nullable(&self) -> bool {
+        self.ty.nullable
+    }
+
+    /// Returns an [`InnerType`] which represents the inner value of the type.
+    /// If the type is a list, the inner type is the type of the list's elements.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::{*};
+    ///
+    /// let individual_ty = Type::new("Int!").unwrap();
+    /// assert!(matches!(individual_ty.value(), InnerType::NameOfType("Int")));
+    ///
+    /// let list_ty = Type::new("[Int!]!").unwrap();
+    /// let inner_ty_for_assert = Type::new("Int!").unwrap();
+    /// assert!(matches!(list_ty.value(), InnerType::ListInnerType(inner_ty) if inner_ty == inner_ty_for_assert));
+    /// ```
+    pub fn value(&self) -> InnerType<'_> {
```

Comment:
Is `value` a bad name for this? Another candidate for this could be `inner_value`. Or, we could separate out what this enum gives us to two separate methods but I think it's much more useful and idiomatic this way, despite us doing more work in some cases to clone the List's inner type for the List() even if that branch is ignored. (although maybe that's optimized away? not sure)

Reply:
Yeah, it's not a great name.

I'm also really not sure about `InnerType` as a whole. When you find it difficult to come up with a good name for a concept, oftentimes it's an indicator of a confusing or muddled design.

Reply:
Did you have a better interface in mind for this? I think that this is the most ergonomic way to expose the data that the user wants to know of is this a NameOfType (Named Type) or a ListInnerType (List). For the most part this is what we already have with the BaseType enum, so it doesn’t require too much code changes, but if you have a better design you can see, please share.

Reply:
It would require a lot of testing, but if you're up for it, here's a fairly compact and efficient `Type` implementation with no pointer chasing in the common path, and with all the internals hidden:
```rust
pub struct Type {
    base: Arc<str>,
    modifiers: Modifiers,
}

// This type is meant to be tiny: discriminant + 1 pointer/u64 union
enum Modifiers {
    Inline(InlineModifiers),
    Extended(Box<ExtendedModifiers>),
}

struct InlineModifiers {
    mask: u64,  // space for ~30 levels of list nesting
}

impl InlineModifiers {
    const NON_NULLABLE_MASK: u64 = 1;
    const LIST_MASK: u64 = 2;

    fn is_nullable(&self) -> bool {
        (self.mask & Self::NON_NULLABLE_MASK) == 0
    }

    fn is_list(&self) -> bool {
        (self.mask & Self::LIST_MASK) != 0
    }

    fn as_list(&self) -> Option<InlineModifiers> {
        self.is_list().then_some(|| {
            InlineModifiers::new(self.mask >> 2)
        })
    }
}

// In case someone really wants more than ~30 levels of nested lists in their type...
struct ExtendedModifiers {
    // ... TODO ...
}

impl Type {
    pub fn nullable(&self) -> bool {
        self.modifiers.is_nullable()
    }

    pub fn is_list(&self) -> bool {
        self.modifiers.is_list()
    }

    pub fn list_of(&self) -> Option<Self> {
        self.modifiers.as_list().map(|modifiers| {
            Self {
                base: self.base.clone(),
                modifiers,
            }
        })
    }

    pub fn base_type(&self) -> &str {  // this ensures the Arc<str> is an implementation detail
        self.base.as_ref()
    }
}
```

Reply:
@obi1kenobi I think that initial definition of `Type` is a good start. If we want to optimize even further, there's some options to reduce `Type` from 4 words. These are rather tricky to get right, and we can optimize piecemeal, so **I recommend against doing any of these in this commit**.

To preface, we have an abundance of free real estate for tags:
- The most significant bit of `str` length because it cannot exceed `isize::MAX`
- The bottom bits of pointers because we can enforce the alignment. For example:
    - 2 byte align: 1 bit
    - 4 byte align: 2 bits
    - 8 byte align: 3 bits

Given the space for tags, we have various options for optimizing `Type`:
- Make `Type` 3 words by making `Modifiers` a tagged pointer, leaving us 63 bits for `InlineModifiers`
- Make `Type` 2 words by storing `ExtendedModifiers` as the header for the base name `str`
- Make `Type` 1 word by moving `str` length into the allocation along with the optional `ExtendedModifiers` header
- Compile-time types!
    - Tag `Type` to indicate `&'static` instead of `Arc`
    - Allows commonly-known types to have free `Clone` and no heap alloc

Reply:
Great to know we'll be set up for future optimizations, should they prove necessary.

Let's start with the initial `Type` layout (which should already eliminate a lot of pointer-chasing and be faster for it), and we can optimize further in the future — which won't even be a breaking change from here onward!

@u9g do you feel like you can take this initial sketch of a design for `Type` and run with it from here?

Reply:
Bonus optimization: `Type` could also store tiny strings inline, up to 7 bytes for 1 word `Type`

Reply:
Yes, I'll start working with the initial design and we can optimize it later :)

Reply:
The initial implementation has landed!